### PR TITLE
IDGEN-42: Add webservices to idgen module

### PR DESCRIPTION
**Pull request by samuelmale**
Ticket : https://issues.openmrs.org/browse/IDGEN-42

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/idgen/SequentialIdentifierGenerator.java [hunk 1]
```diff
@@ -177,12 +177,12 @@ public void setBaseCharacterSet(String baseCharacterSet) {
 	/**
 	 * @return the nextSequenceValue
 	 */
-	public Long getNextSequenceValue() {
-		if(nextSequenceValue == null) return -1l;
```
[the remark this answers is not in the document]
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/idgen/SequentialIdentifierGenerator.java]: Gonna revert that @dkayiwa . Just go on and make more objections

File: api/src/test/java/org/openmrs/module/idgen/SequentialIdentifierGeneratorTest.java [hunk 2]
```diff
@@ -62,15 +62,4 @@ public void getIdentifierForSeed_shouldThrowAnErrorIfGeneratedIdentifierIsLonger
 		generator.getIdentifierForSeed(1);
 	}
 
-	@Test
-	public void shouldSetNextSequenceValueToNegative() throws Exception {
-		SequentialIdentifierGenerator generator = new SequentialIdentifierGenerator();
-		generator.setBaseCharacterSet("0123456789");
-		generator.setPrefix("FOO-");
```
Comment (dkayiwa) [inline api/src/test/java/org/openmrs/module/idgen/SequentialIdentifierGeneratorTest.java]: Why did we remove this test?

Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/idgen/SequentialIdentifierGeneratorTest.java]: I'm gonna revert this also :)

File: omod/src/test/java/org/openmrs/module/idgen/rest/resource/handler/IdentifierPoolResourceHandlerTest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	IdentifierSourceService service;
+	
+	@Autowired
+	private IdentifierPoolResourceHandler handler;
+	
+	private IdentifierPool pool;
+	
+	@Before
+	public void setup() {
+		pool = new IdentifierPool();
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/rest/resource/handler/IdentifierPoolResourceHandlerTest.java]: Why are you doing this from here?

File: omod/src/test/java/org/openmrs/module/idgen/rest/resource/handler/IdentifierPoolResourceHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+public class IdentifierPoolResourceHandlerTest extends BaseModuleWebContextSensitiveTest {
+	
+	@Autowired
+	IdentifierSourceService service;
+	
+	@Autowired
+	private IdentifierPoolResourceHandler handler;
+	
+	private IdentifierPool pool;
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/rest/resource/handler/IdentifierPoolResourceHandlerTest.java]: Why is this a class level variable?

File: omod/src/test/java/org/openmrs/module/idgen/rest/resource/handler/SequentialIdentifierGeneratorResourceHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	
+	private SequentialIdentifierGenerator generator;
+	
+	private DelegatingResourceDescription representationDescription;
+	
+	private Representation rep;
+	
+	@Before
+	public void setup() {
+		generator = new SequentialIdentifierGenerator();
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/rest/resource/handler/SequentialIdentifierGeneratorResourceHandlerTest.java]: Why are you setting these from here? Why are they class level variables?

**Comment by dkayiwa** [comment]
Can you also make these resources show up here https://demo.openmrs.org/openmrs/module/webservices/rest/apiDocs.htm by implementing these methods? https://wiki.openmrs.org/display/docs/Documenting+REST+Resources

**Comment by tendomart** [comment]
@dkayiwa  for the addition, had not considered the fact the this should actually documented against Swagger UI generator.

**Comment by samuelmale** [comment]
@dkayiwa can you now have a look?

**Comment by dkayiwa** [comment]
Did you compile and install the module to confirm that its swagger documentation is also listed?

**Comment by dkayiwa** [comment]
@samuelmale ping

**Comment by samuelmale** [comment]
Well I tested it out but it didn't show the docs but am debugging it to prove why

File: api/src/main/java/org/openmrs/module/idgen/Identifier.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+/**
+ * This is a dummy wrapper class for an identifier {@link String}
+ * It's required for the {@link IdentifierResource} to compile when extending {@link DelegatingSubResource} as required to include 
+ * the {@link IdentifierResource} in the Swagger docs
+ * 
+ * @since 4.6.0
+ */
+public class Identifier {
+
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/idgen/Identifier.java]: What is this class for?

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java [hunk 7]
```diff
@@ -10,16 +10,21 @@
...
+import org.openmrs.module.webservices.rest.web.resource.impl.DelegatingResourceDescription;
+import org.openmrs.module.webservices.rest.web.resource.impl.DelegatingSubResource;
 import org.openmrs.module.webservices.rest.web.response.ResponseException;
 
 @org.openmrs.module.webservices.rest.web.annotation.SubResource(parent = IdentifierSourceResource.class, path =
-        "identifier", supportedClass = String.class, supportedOpenmrsVersions = {"1.8.*", "1.9.*", "1.10.*", "1.11.*", 
+        "identifier", supportedClass = Identifier.class, supportedOpenmrsVersions = {"1.8.*", "1.9.*", "1.10.*", "1.11.*", 
         		"1.12.*", "2.0.*", "2.1.*", "2.2.*" })
-public class IdentifierResource implements SubResource {
+public class IdentifierResource extends DelegatingSubResource<Identifier, IdentifierSource, IdentifierSourceResource> {
```
Comment (samuelmale) [inline omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: @dkayiwa  if we are to add the IdentifierResource to the docs, we need to extend the BaseDelegatingResource inorder to override the required methods, that being said the IdentifierResource was supporting the String class(identifier string) incase we extend the BaseDelegatingResource class with the String class as the type we are delegating to, we shall get duplicate methods here https://github.com/openmrs/openmrs-module-webservices.rest/blob/master/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/impl/BaseDelegatingResource.java#L544-L566

Comment (samuelmale) [reply omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Hence hitting compiling errors

File: api/src/main/java/org/openmrs/module/idgen/Identifier.java (hunk 6 above)
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/idgen/Identifier.java]: @dkayiwa  if we are to add the IdentifierResource to the docs, we need to extend the BaseDelegatingResource inorder to override the required methods, that being said the IdentifierResource was supporting the String class(identifier string) incase we extend the BaseDelegatingResource class with the String class as the type we are delegating to, we shall get duplicate methods here https://github.com/openmrs/openmrs-module-webservices.rest/blob/master/omod-common/src/main/java/org/openmrs/module/webservices/rest/web/resource/impl/BaseDelegatingResource.java#L544-L566

**Comment by dkayiwa** [comment]
@samuelmale which version of the OpenMRS platform are you using to test this?

**Comment by dkayiwa** [comment]
Are all the resources supporting 2.2.*?

**Comment by samuelmale** [comment]
Yes, currently all of them

**Comment by dkayiwa** [comment]
Are you sure?

**Comment by samuelmale** [comment]
Pretty sure!

**Comment by samuelmale** [comment]
Did you agree with the reason for need of a new (wrapper)Class for Identifiers?

**Comment by dkayiwa** [comment]
Not yet convinced.

**Comment by dkayiwa** [comment]
Did you also see this error in the generated docs? Could not resolve reference because of: Could not resolve pointer: /definitions/IdgenIdentifiersourceIdentifierCreate does not exist in document

**Comment by samuelmale** [comment]
> Did you also see this error in the generated docs? Could not resolve reference because of: Could not resolve pointer: /definitions/IdgenIdentifiersourceIdentifierCreate does not exist in document

I don't get that error at my end. All the representations for the IdentifierSourceResource are well documented and they work well unlike its subclasses

> Not yet convinced.

1. Did you see my explanation as to why I created a new class?
2. I believe it's you that added the IdentifierResource, why did you just implement `SubResource` rather than extending `DelegatingSubResource`? You will find it's not possible to extend `DelegatingSubResource` with the type of Resource we are delegating to as `String`.

**Comment by dkayiwa** [comment]
I do not see the t parameter on the idgen identifiersource resource, even after running these changes. I only see it on the order resource.

**Comment by samuelmale** [comment]
Gonna push some updates to resolve this

**Comment by dkayiwa** [comment]
Looking forward to it.

File: api/src/main/java/org/openmrs/module/idgen/IdentifierPool.java [hunk 8]
```diff
@@ -131,6 +133,7 @@ public void setMinPoolSize(Integer minPoolSize) {
 	/**
 	 * @return boolean describing whether this is sequential
 	 */
+	@JsonIgnore
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/module/idgen/IdentifierPool.java]: Why ignore this?

Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/idgen/IdentifierPool.java]: As a matter of fact that we have #getSequential() which the standard getter, while Jackson is parsing this file into json it also evaluates this as a getter hence resolving into conflicting getters

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierSourceResource.java [hunk 9]
```diff
@@ -493,6 +503,39 @@ protected PageableResult doSearch(RequestContext context) throws ResponseExcepti
         return new EmptySearchResult();
     }
     
+    @Override
+    public boolean hasTypesDefined() {
+        return true;
```
Comment (samuelmale) [inline omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierSourceResource.java]: Overriding this has influenced lots of changes in this PR. Especially those specific to Jackson eg. @JsonIgnore annotations

**Comment by dkayiwa** [comment]
Do you also get these errors?

Resolver error at paths./idgen/identifiersource/{parent-uuid}/identifier.get.responses.200.schema.properties.results.items.$ref
Could not resolve reference because of: Could not resolve pointer: /definitions/IdgenIdentifiersourceIdentifierGet does not exist in document
Resolver error at paths./idgen/identifiersource/{parent-uuid}/identifier/{uuid}.get.responses.200.schema.$ref
Could not resolve reference because of: Could not resolve pointer: /definitions/IdgenIdentifiersourceIdentifierGet does not exist in document

**Comment by samuelmale** [comment]
Yes I also do but I wonder why. But I believe those errors are specific to
webservices.rest module the fix should be applied their not in my
implementation. Do they block IDGEN-42?

**Comment by dkayiwa** [comment]
Do you know why resources like logentry, autogenerationoption, are not listed in the documentation?

**Comment by samuelmale** [comment]
Where are those resources defined from, a module?

**Comment by dkayiwa** [comment]
https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/main/java/org/openmrs/module/idgen/rest/resource/AutoGenerationOptionResource.java

https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/main/java/org/openmrs/module/idgen/rest/resource/LogEntryResource.java

**Comment by samuelmale** [comment]
I have added support for swagger documentation of those two resources. Thanks for spotting them.

**Comment by dkayiwa** [comment]
Do you think you can also deal with the above two swagger doc errors from the idgen module?

**Comment by dkayiwa** [comment]
Did you also notice that your followup commit introduced more swagger doc errors?

**Comment by samuelmale** [comment]
Which commit?

**Comment by samuelmale** [comment]
Hope we are good. Do you still get errors in the doc?

**Comment by dkayiwa** [comment]
I mean the commit you made yesterday to address the two missing resources.
Do you mean you do not have any swagger errors your side?

**Comment by samuelmale** [comment]
I have solved everything now. Are you able to test out it now?

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java [hunk 10]
```diff
@@ -61,11 +87,56 @@ public void purge(String parentUniqueId, String uniqueId, RequestContext context
...
+	}
+
+	@Override
+	public void purge(Identifier delegate, RequestContext context) throws ResponseException {
+		throw new ResourceDoesNotSupportOperationException();		
+	}
+
+	@Override
+	public DelegatingResourceDescription getRepresentationDescription(Representation rep) {
+		throw new UnsupportedOperationException();
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Are you sure we do not support all these operations?

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java [hunk 11]
```diff
@@ -141,8 +140,8 @@ public void shouldUploadIdentifiersFromSource() throws Exception {
         SimpleObject identifiersAfterUpload = deserialize(handle(getRequestAfterUpload));
         assertThat(
                 identifiersAfterUpload.get("identifiers").toString(), allOf(
-                containsString("Test Identifier Pool: G-0"),
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java]: Why did you change these?

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java [hunk 12]
```diff
@@ -164,10 +163,10 @@ public void shouldUploadIdentifiersFromFile() throws Exception {
         SimpleObject identifiersAfterUpload = deserialize(handle(getRequestAfterUpload));
         assertThat(
                 identifiersAfterUpload.get("identifiers").toString(), allOf(
-                containsString("Test Identifier Pool: 1"),
-                containsString("Test Identifier Pool: 2"),
```
Comment (dkayiwa) [inline omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java]: Why did you change these?

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java (hunk 10 above)
Comment (samuelmale) [reply omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Yes it's trivial overriding this since we don't support any get operations. We only support POST for generating Identifiers. 
BTW, whats the significancy of overriding this method, just to insight me!

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java (hunk 11 above)
Comment (samuelmale) [reply omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java]: It's because of this change (https://github.com/openmrs/openmrs-module-idgen/pull/74/files#diff-37ab2fee2e4e68dcf200614c09c4a51dR24 -PooledIdentifier). I removed the the IdentifierPool from PooledIdentifier model since its a cyclic dependency and could cause serialization issues with Jackson.

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java [hunk 13]
```diff
@@ -102,7 +102,7 @@ public Identifier newDelegate() {
 
 	@Override
 	public Identifier save(Identifier delegate) {
-		throw new ResourceDoesNotSupportOperationException();
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Any reason why you are some times using ResourceDoesNotSupportOperationException and other times UnsupportedOperationException?

Comment (samuelmale) [reply omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Well your right on this. This is a tweak or hack around to make the POST path for this resource be added in the docs. Well look at this [method that determines supported operations](https://github.com/openmrs/openmrs-module-webservices.rest/blob/master/omod-common/src/main/java/org/openmrs/module/webservices/docs/swagger/SwaggerSpecificationCreator.java#L331). I assumes a Resource that supports `#create()` should also support `#save()` operations which isn't the case for this resource. That's why I intended to throw UnsupportedOperationException which is an instance of ResponseException but is not checked by the `#testOperationImplemented()` in the webservices.rest module. This way the POST path can be included in the docs. The other alternative was to just `return null;` but I found it trivial

**Comment by samuelmale** [comment]
ping @dkayiwa

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java (hunk 13 above)
Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Doesn't it make more sense to instead fix testOperationImplemented()?

Comment (samuelmale) [reply omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Well it makes more sense and meaning but I believe that is out of scope for this ticket. I suppose it shouldn't block closure of this issue just according to our RoadMap  :)

Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Why would we hack this module instead of just fixing the other?

**Comment by samuelmale** [comment]
However for the fix of the hack to work we need to depend of the latest changes in webservice.rest module.

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java (hunk 12 above)
Comment (samuelmale) [reply omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java]: It's because of this change (https://github.com/openmrs/openmrs-module-idgen/pull/74/files#diff-37ab2fee2e4e68dcf200614c09c4a51dR24 -PooledIdentifier). I removed the the IdentifierPool from PooledIdentifier model since its a cyclic dependency and could cause serialization issues with Jackson.

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java [hunk 14]
```diff
@@ -30,14 +40,33 @@ public String getUri(Object instance) {
 
     @Override
     public Object create(String parentUniqueId, SimpleObject post, RequestContext context) throws ResponseException {
+    	if (post == null) {
+    		return null;
+    	}
         IdentifierSourceService service = Context.getService(IdentifierSourceService.class);
         String identifier = service.generateIdentifier(service.getIdentifierSourceByUuid(parentUniqueId),
-                post.containsKey("comment") ? String.valueOf(post.get("comment")) : "");
+                post.containsKey("comment") ? String.valueOf((String)post.get("comment")) : "");
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Do you really need casting this to a String?

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java (hunk 10 above)
Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/idgen/rest/resource/IdentifierResource.java]: Any reason for mixing ResourceDoesNotSupportOperationException and UnsupportedOperationException?

File: omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java (hunk 12 above)
Comment (dkayiwa) [reply omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java]: I still do not understand how that makes it change from "Test Identifier Pool: 2" to "2"

Comment (samuelmale) [reply omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java]: Since I removed the IdentifierPool obviously its name property won't be included in the String we are asserting.

Still curious? See [this](https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java#L162), references [this](https://github.com/openmrs/openmrs-module-idgen/blob/master/omod/src/test/java/org/openmrs/module/idgen/web/controller/IdentifierSourceRestControllerTest.java#L43) which later references [this](https://github.com/openmrs/openmrs-module-idgen/blob/master/api/src/test/resources/org/openmrs/module/idgen/include/TestData.xml#L5) - I believe you have seen the name property "Test Identifier Pool"!

File: omod/src/main/java/org/openmrs/module/idgen/rest/resource/handler/IdentifierPoolResourceHandler.java [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+					.property("sequential", new BooleanProperty())
+					.property("refillWithScheduledTask", new BooleanProperty())
+					.property("source", new RefProperty("#/definitions/IdentifierSourceGet"))
+					.property("batchSize", new IntegerProperty())
+					.property("minPoolSize", new IntegerProperty());
+	}
+	
+	@Override
+	public Model getUPDATEModel(Representation rep) {
+		return getCREATEModel(rep); //FIXME add Impl
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/idgen/rest/resource/handler/IdentifierPoolResourceHandler.java]: When do you plan to fix the above?
